File: servers/js/mtls_server.js
const tls = require("tls");
const fs = require("fs");

const CERT_DIR = `/certs/ml-dsa-${process.env.ML_DSA_LEVEL || "44"}`;
const PORT = 8444;

const server = tls.createServer(
  {
    cert: fs.readFileSync(`${CERT_DIR}/server.pem`),
    key: fs.readFileSync(`${CERT_DIR}/server.key`),
    ca: fs.readFileSync(`${CERT_DIR}/ca-chain.pem`),
    requestCert: true,
    rejectUnauthorized: true,
    minVersion: "TLSv1.3",
  },
  (socket) => {
    const client = socket.getPeerCertificate();
    console.log(
      `handshake OK with ${socket.remoteAddress}:${socket.remotePort}: ${socket.getProtocol()} ${socket.getCipher().standardName} client_cn=${client.subject && client.subject.CN}`
    );
    socket.on("data", () => {
      socket.end("HTTP/1.1 200 OK\r\nContent-Length: 2\r\nConnection: close\r\n\r\nok");
    });
  }
);

server.on("tlsClientError", (err, socket) => {
  console.error(`connection FAILED with ${socket.remoteAddress}:${socket.remotePort}: ${err.message}`);
});

server.listen(PORT, "0.0.0.0", () => {
  console.log(`mTLS server (client cert required) listening on :${PORT}`);
});

File: servers/python/mtls_server.py
import os
import ssl
import socket
import sys

CERT_DIR = f"/certs/ml-dsa-{os.environ.get('ML_DSA_LEVEL', '44')}"
PORT = 8444


def main():
    ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
    ctx.minimum_version = ssl.TLSVersion.TLSv1_3
    ctx.load_cert_chain(certfile=f"{CERT_DIR}/server.pem", keyfile=f"{CERT_DIR}/server.key")
    ctx.verify_mode = ssl.CERT_REQUIRED
    ctx.load_verify_locations(cafile=f"{CERT_DIR}/ca-chain.pem")

    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.bind(("0.0.0.0", PORT))
        sock.listen(5)
        print(f"mTLS server (client cert required) listening on :{PORT}", flush=True)
        while True:
            conn, addr = sock.accept()
            try:
                with ctx.wrap_socket(conn, server_side=True) as ssock:
                    peer = ssock.getpeercert()
                    subject = dict(x[0] for x in peer.get("subject", [])) if peer else {}
                    print(f"handshake OK with {addr}: {ssock.version()} {ssock.cipher()} client_cn={subject.get('commonName')}", flush=True)
                    try:
                        ssock.recv(4096)
                    except Exception:
                        pass
                    ssock.sendall(b"HTTP/1.1 200 OK\r\nContent-Length: 2\r\nConnection: close\r\n\r\nok")
            except (ssl.SSLError, OSError) as e:
                print(f"connection FAILED with {addr}: {e}", file=sys.stderr, flush=True)


if __name__ == "__main__":
    main()
